Zend/zend_closures.c

```diff
@@ -124,17 +115,23 @@ ZEND_METHOD(Closure, call)
 
 	if (fci_cache.function_handler->common.fn_flags & ZEND_ACC_GENERATOR) {
 		zval new_closure;
-		zend_create_closure(&new_closure, fci_cache.function_handler, Z_OBJCE_P(newthis), closure->called_scope, newthis);
+		zend_create_closure(&new_closure, fci_cache.function_handler, closure->func.common.scope, closure->called_scope, &closure->this_ptr);
 		closure = (zend_closure *) Z_OBJ(new_closure);
+		closure->func.common.fn_flags |= ZEND_ACC_PUBLIC;
+		zval_ptr_dtor_nogc(&closure->this_ptr);
+		ZVAL_COPY(&closure->this_ptr, newthis);
 		fci_cache.function_handler = &closure->func;
 	} else {
 		memcpy(&my_function, fci_cache.function_handler, fci_cache.function_handler->type == ZEND_USER_FUNCTION ? sizeof(zend_op_array) : sizeof(zend_internal_function));
-		/* use scope of passed object */
-		my_function.common.scope = Z_OBJCE_P(newthis);
 		fci_cache.function_handler = &my_function;
+	}
+
+	if (closure->func.type == ZEND_USER_FUNCTION && (closure->func.common.fn_flags & ZEND_ACC_REAL_CLOSURE)) {
+		/* use scope of passed object; we must not change scope of functions and methods, only true Closures */
+		fci_cache.function_handler->common.scope = Z_OBJCE_P(newthis);
```

Comment:
That will though have the side effect that the set of possible Closures you actually can pass to a function, thus rendering the benefits of Closure::call() minimal, I think.

Also, totally forgot, Closure::call() is anyway new in PHP 7…